### ISSUE TITLE
Always getting "No response from query server" behind NAT

### ISSUE THREAD

**Issue by alfonx** [OmniDB/OmniDB#342]
I installed OmniDB 2.4.1 on Ubuntu 16.04 using .deb on Server A, and I configured it to access PG10 on Server B in the local network.
I can log in, create a connection, it shows a dashboard with data, but executing ANY times out with "No response from query server". Nothing in the serverlog, nor in the ~/.omnidb/omnidb-server/omnidb.log

I have no clue how to debug this. Any ways to increase omnidb log level?

Is the installation of the plugin (https://github.com/OmniDB/OmniDB/blob/master/omnidb_plugin/README.md) mandatory?

**Comment by rafaelthca** [comment]
Hi Stefan,

The plugin is not mandatory, only if you are debugging functions.

I see you are using omnidb-server, right? What browser are you using? What version?

Interface is probably not connected to websocket server, if you can please open developer tools (F12) and check if there are any errors in the console tab.

For a more detailed description of the error go to the Network tab in developer tools and filter WS. Then refresh the page to record the browser trying to connect to the websocket server.

**Comment by alfonx** [comment]
I use FF 57. You are right! omnidb-server runs on Port 8000 and is reachable from the outside-world via NAT on Port 9999:

>Firefox kann keine Verbindung zu dem Server unter wss://my.external.dns:8000/ws aufbauen.

Seems like it confuses the local port 8000 and the external NATed port 9999 for the websocket. I guess I can help myself by setting both to the same port....

**Comment by alfonx** [comment]
I have set the external NAT-Port equal to the local port of the omnidb-server.

The can't connect network-error "Firefox kann keine Verbindung zu dem Server unter wss://my.external.dns:8000/ws aufbauen." is gone.

But it still doesn't work. FF not reporting an error in the console anymore, BUT I see an empty response after timeout to wss://my.external.dns:9999/ws now..

The outcome is the same: "No response from query server"

The firewall is NATting TCP on Port 9999 to omnidb-server. Anything else needed? UDP?

**Comment by rafaelthca** [comment]
What is the status code of the ws connection?

![image](https://user-images.githubusercontent.com/9335851/35225747-1ff7b3f4-ff57-11e7-94f4-cc69b4cefb59.png)

**Comment by alfonx** [comment]
It was 101 Switching Protocols when I tried it bevore.

Anyways.. I was using Ctrl-Shift-r to empty caches all the time testing.. Now I restarted Firefox and it works!

Feel free to close the ticket and Thanks a lot for your help!

Probably it should be mentioned somewhere that outside NATed port and local port must be the same.

**Comment by rafaelthca** [comment]
This is actually a bug, client should use the same port being used by the url in the browser, not the port OmniDB is actually running, I will fix this. Thanks for reporting.

Keeping this open until bug is fixed.

**Comment by wind39** [comment]
Hi @alfonx ,

Can you please try release 2.5.0 we published today? This bug should be fixed, allowing you to specify different NAT and server ports.

Regards,
William

**Comment by deric** [comment]
I'm having similar issue (btw. there's no release 2.5.0 on github).

```
WebSocket connection to 'wss://omnidb.example.net:25482/wss' failed: WebSocket opening handshake timed out
``` 
I'd expect that wss would open either on port 80 or 443. Obviously port `25482` is blocked on the firewall.

**Comment by deric** [comment]
The problem is [here](https://github.com/OmniDB/OmniDB/blob/master/OmniDB/OmniDB_app/views/workspace.py#L45):
```
        'query_port': settings.OMNIDB_PORT,
```
the query port is set to a e.g. private port inside docker container which doesn't correspond to load balancer's port. Moreover it looks like the ENV variable [is overridden by a default value](https://github.com/OmniDB/OmniDB/blob/master/OmniDB/OmniDB/settings.py#L218):
```
OMNIDB_PORT                = 25482
```

**Comment by wind39** [comment]
Hi @deric ,

Can you please test it on 2.5.0? The release is here: https://github.com/OmniDB/OmniDB/releases/tag/2.5.0

**Comment by deric** [comment]
@wind39 Yes, I'm running on the latest version in master branch 8de6dbb47bbdb028671cf0ea79550f902e6162fd which appears to be the same.

**Comment by deric** [comment]
I've switched the port to `443` and now I'm getting 
```
WebSocket connection to 'wss://omnidb.example.net/wss' failed: Error during WebSocket handshake: Unexpected response code: 404
```

**Comment by rafaelthca** [comment]
Hi @deric,

So, I'm assuming you're running omnidb in your python environment (not using omnidb-server), is that correct?

If so, can you please specify how the django app being served? Also, what load balancer is being used?

**Comment by rafaelthca** [comment]
I was able to reproduce your issue @deric, here is the problem: OmniDB now uses tornado to serve both http/https and ws/wss requests

python manage.py runserver (or other application serving omnidb) will conflict with tornado and you won't be able to make it work properly, it is used only for development purposes.

Please use 'python omnidb-server.py' to start your server, that setting OMNIDB_PORT will be overwritten accordingly.

**Comment by rafaelthca** [comment]
Another point, settings.py has this parameter: DEV_MODE = True. Please set it to False if you want to use OmniDB behind a load balancer.

**Comment by deric** [comment]
@rafaelthca I'm using nginx with SSL certificates that normally proxies requests to OmniDB backend. Nothing unusual. `omnidb-server.py` has hardcoded workdir, it's not realy compatible with the provided `Dockerfile`:
```
workdir = '/opt/omnidb-server/'
```
The web interface is somehow working but no SQL query can be executed.

**Comment by rafaelthca** [comment]
Hi Deric, I'm sorry you're having trouble to set up your environment. You are able to access omnidb and use the interface but not query anything because omnidb uses a separate websocket server to perform the queries, to enable bidirectional communication in the interface. When you run omnidb with the packages we provide it uses the same port for both http/https requests and ws/wss requests. If you run omnidb from its source code with python manage.py runserver it will use different ports (the port you specified in runserver will serve http and that 25482 will serve ws requests), but this mode is used only for development.

You're right, omnidb-server.py isn't prepared to be executed if not using our Packages that are available for download. I created a Dockerfile that uses debian stable slim and installs the omnidb-server package, if you can please test it:

```
FROM debian:stable-slim

RUN  apt-get update \
  && apt-get install -y wget \
  && rm -rf /var/lib/apt/lists/*

RUN mkdir /app
WORKDIR /app

RUN wget https://omnidb.org/dist/2.5.0/omnidb-server_2.5.0-debian-amd64.deb

RUN dpkg -i /app/omnidb-server_2.5.0-debian-amd64.deb

EXPOSE 8000

CMD ["omnidb-server"]
```
```
docker build -t omnidb-server .
docker run -d -p 12000:8000 omnidb-server
```
Then just forwards requests in nginx to 12000.

I actually use omnidb-server in my cloud machine behind NGINX, it requires additional configuration to proxy WS requests.

```
location / {
      proxy_pass https://127.0.0.1:10551;

      proxy_set_header   X-Real-IP $remote_addr;
      proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header   X-Forwarded-Ssl https;
      proxy_set_header   X-Forwarded-Proto https;
      proxy_set_header   X-Forwarded-Port 443;
      proxy_set_header   Host $host;

      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
}
```

This is my setup, but omnidb-server is not in a docker container, it is running in the same machine as NGINX, running on port 10551 with SSL mode set to true, using the same certificates used by NGINX

**Comment by wind39** [comment]
Hi @alfonx and @deric ,

We implemented several improvements on this. @rafaelthca wrote a chapter in OmniDB docs [1] to explain how to deploy omnidb-server in different scenarios. Can you please take a look? We'd like to have a feedback on this, please, so we are keeping this open for now.

Thanks and regards,
William Ivanski

[1] https://omnidb.org/en/documentation-en/19-deploying-omnidb-server

**Comment by alfonx** [comment]
Thanks for working on this!

After upgrading to omnidb 2.6.0 and reading the new documentation I cant get it to work anymore, so I go back one step back to where external ports and local ports are equal...

NAT is now configured most simple:
>WAN-IP:9999 and :10000 to LAN-IP:9999 and :10000

The omnidb.conf defines:
>listening_address    = 0.0.0.0
>listening_port       = 9999
>websocket_port       = 10000
>is_ssl               = True
>ssl_certificate_file = /etc/ssl/certs/my.crt
>ssl_key_file         = /etc/ssl/private/my.key

With this setup I can login, but I can not do a query.

And the webinterface complains:
>Cannot connect to websockes server with ports 10000 (external) and 10000 (internal).

The omnidb-server compains:
>Starting OmniDB websocket...
>Checking port availability...
>Starting websocket server at port 10000.
>Starting OmniDB server...
>Checking port availability...
>Starting server OmniDB 2.6.0 at 0.0.0.0:9999.
>WARNING:tornado.general:error on read: [SSL: SSLV3_ALERT_BAD_CERTIFICATE] sslv3 alert bad certificate (_ssl.c:1977)


OK.. I agree that the certificate does not match the local ip adress, but I still want the certificate to be used, and it worked in 2.5.0.

So my NAT setup again:

I want omnidb to run HTTPS one a mashine in the local network.
I want to tell my firewall using NAT that this service is reachable from outside.
This setup was working in 2.5.0. Could there be a regression?

Thanks and regards,
Stefan

**Comment by rafaelthca** [comment]
Hi @alfonx,

Thank you again for reaching out. The omnidb.conf file (starting in 2.6.0) comes with the additional parameter:

# External Websocket port, use this parameter if OmniDB isn't directly visible by the client
# external_websocket_port = 25482

That can be used when OmniDB isn't directly accessed, which is your case behind the firewall. The parameter can be set to a value different from websocket_port. The interface will then try to connect to both ports, since you didn't specify the external one OmniDB is using the same value for both external and internal.

Your problem seems to be other in this case. Are you saying that OmniDB was initiated with SSL set to true, passing a certificate of a specific domain and then you are trying to connect to OmniDB locally? If done like that the server should complain about domain mismatch, was it working on 2.5.0?

Can you please share again screenshots of your browser developer tools so I can see the status of the websocket connection?

**Comment by alfonx** [comment]
>  since you didn't specify the external one OmniDB is using the same value for both external and internal.
Yes, that is the simple case I try to get running first now.

NAT is mapping 9999->9999 and 10000->10000

So I don't think I nedd to set external_websocket_port to 10000

> Your problem seems to be other in this case. Are you saying that OmniDB was initiated with SSL set to >true, passing a certificate of a specific domain and then you are trying to connect to OmniDB locally? If >done like that the server should complain about domain mismatch, was it working on 2.5.0?
Yes that is what I want to do, and it worked in 2.5.0.
Of course the CLIENT complained that the domain missmatched, but should the Server really complain?

Has OmniDB 2.6.0 websocket server changed? Does it check for every request if the domain I requested in the client fits the domain? (I am pretty sure old friends like appache never do this check and leave it to the client)

So, do I have to create a new cerfiticate only for OmniDB 2.6.0 ?

Thanks

**Comment by alfonx** [comment]
When I disable is_ssl it is working (NAT is mapping 9999->9999 and 10000->10000, external_websocket_port not set). 
So I can't really say wether the behind-NAT-problem is solved with "external_websocket_port".

OmniDB wihtout encryption is my bigger problem now ;-) ... Maybe worth a new ticket?

**Comment by rafaelthca** [comment]
> Has OmniDB 2.6.0 websocket server changed? Does it check for every request if the domain I requested in the client fits the domain? (I am pretty sure old friends like appache never do this check and leave it to the client)

We didn't change that, but 2.6.0 is using a different version of the websocket server. I tried to simulate your scenario since you mentioned that 2.5.0 was working fine (can you confirm that?) but even older versions of the websocket server validate domains when ssl is enabled.

I am thinking about including an option to keep using SSL and disable domain validation.

> OmniDB wihtout encryption is my bigger problem now ;-)

Right, encryption is recommended. There are several approaches as a workaround for your scenario:

1- Use omnidb without SSL listening only in the local network and behind NGINX. Local access to omnidb will not be encrypted. NGINX handles external requests with encryption.

2- If you need ssl even in the local network. Start OmniDB without encryption listening on 127.0.0.1 and use NGINX to handle external requests with a domain certificate and internal requests with a self signed certificate.

3- (Ugly one) Hack your /etc/hosts to resolve the required domain to 127.0.0.1 or whatever address OmniDB is running on.

I will investigate this further and would really appreciate if you can share more details about your scenario so I can improve this part.

New version is coming out in around 3 weeks and we can possibly include a fix for that.

So, can you confirm it was working in 2.5.0?

**Comment by alfonx** [comment]
Thanks for you quick feedback!

>So, can you confirm it was working in 2.5.0?
Yes, 100% sure i had not changed the three SSL-related lines in omnidb.conf from 2.5.0 to 2.6.0 and it was working in 2.5.0 (with a SSL domain-warning in the browser of course).

>is_ssl               = True
>ssl_certificate_file = /etc/ssl/certs/my.de.crt
>ssl_key_file         = /etc/ssl/private/my.de.key

NGINX is not in my stack on that server, and I don't like to set it up only for omnidb. Of course I could do the same with apache2.. will see how much time I find today.

>WARNING:tornado.general:error on read: [SSL: SSLV3_ALERT_BAD_CERTIFICATE] sslv3 alert bad certificate (_ssl.c:1977)

However, I have one more idea. When I use the certificate in other servers (e.g. apache2) I somethimes have to add one more setting to the config: SSLCertificateChainFile pointing to a second .crt file.
Does omnidb/tornado have anything similar?

BTW: its a wildcard certificate, see it here https://www.ssllabs.com/ssltest/analyze.html?d=www.empirica%2dsysteme.de&latest

**Comment by rafaelthca** [comment]
I did some tests here with my cloud VM. With 2.5.0 and 2.6.0 I achieved the same result, Initiated OmniDB with SSL using a certificate for cloud.omnidb.org (generated with Let's Encrypt). Then tried to access it using the IP instead of the domain. Browser complained of course but I was able to connect to both the interface and to the websocker server, as can be seen here:

![ok_2](https://user-images.githubusercontent.com/9335851/37657664-04139a16-2c22-11e8-9080-8d295228ce23.png)

BTW, what OS is running on your server, and  what OmniDB packages did you use?

Could you share some screenshots of your browser trying to run OmniDB in both 2.5.0 and 2.6.0 so I can see the WS connection status?

Thanks.

**Comment by rafaelthca** [comment]
Finally, was able to simulate your issue. For me it happens with OmniDB 2.6.0 and client accessing with Firefox, with Chrome it works. Can you confirm that? Will continue the investigation

**Comment by alfonx** [comment]
I use https://omnidb.org/dist/2.6.0/omnidb-server_2.6.0-debian-amd64.deb
running on Ubuntu 16.04 LTS

Now i downloaded  https://omnidb.org/dist/2.5.0/omnidb-server_2.5.0-debian-amd64.deb and
did 
```
apt remove omnidb-server
dpkg -i omnidb-server_2.5.0-debian-amd64.deb
```
And omnidb is working again with HTTPS!
![omnidb2 5 0_with_ssl_on_firefox58 0 2](https://user-images.githubusercontent.com/3619833/37658463-7bed1eb8-2c45-11e8-9687-6c813c2b0fce.png)
Changing back to 2.6.0:
```
apt remove omnidb-server
dpkg -i omnidb-server_2.6.0-debian-amd64.deb
```
and it is not working, 


![omnidb2 6 0_with_ssl_on_firefox58 0 2](https://user-images.githubusercontent.com/3619833/37658452-75ab3396-2c45-11e8-90e7-5d85b8ef78cd.png)

with this errors in the server output:
```
Starting OmniDB websocket...
Checking port availability...
Starting websocket server at port 10000.
Starting OmniDB server...
Checking port availability...
Starting server OmniDB 2.6.0 at 0.0.0.0:9999.
User database version 2.6.0 is already matching server version.
Open OmniDB in your favorite browser
Press Ctrl+C to exit
WARNING:tornado.general:error on read: [SSL: SSLV3_ALERT_BAD_CERTIFICATE] sslv3 alert bad certificate (_ssl.c:1977)
WARNING:tornado.general:error on read: [SSL: SSLV3_ALERT_BAD_CERTIFICATE] sslv3 alert bad certificate (_ssl.c:1977)
```

The config used for both is:
```
[webserver]
listening_address    = 0.0.0.0
listening_port       = 9999
websocket_port       = 10000
is_ssl               = True
ssl_certificate_file = /etc/ssl/certs/my.de.crt
ssl_key_file         = /etc/ssl/private/my.de.key
```

I hope this helps a bit...

**Comment by rafaelthca** [comment]
Thank you @alfonx, one last question. What browser (including version) are you using?

**Comment by rafaelthca** [comment]
Ok, I found the issue. From the look of your screenshot you are using Firefox, correct?

When you try to connect to OmniDB using the IP without the domain the browser will complain about about the certificate but you then add the exception and after that you can successfully connect to OmniDB.

Firefox adds an exception for your IP:PORT.

If you go to the firefox settings: Preferences -> Advanced -> Certificates -> View Certificates -> Servers

In that list you will find the exception for the pair (domain / ip:port) that was automatically added when you connected to OmniDB.

Here is the issue: OmniDB 2.5.0 uses the same port to answer omnidb requests and websocket requests. If you check your first screenshot you will see that WS is trying to connect to port 9999 (the same used in the url).

OmniDB 2.6.0 uses different servers for OmniDB and websocket, we did this to improve performance as the other architecture does not perform well when a lot of users are using OmniDB at the same time.

You can see in the second screenshot that the browser is trying to connect to port 10000 in the ws connection.

So, to fix your problem you must manually add an exception to the port 10000. Chrome should work fine.

Let me know how that goes.

**Comment by alfonx** [comment]
You are right: Chromium works with OmniDB 2.6.0

Your are right: I use Firefox 58 ... and just got an update to 59.. both do not work out of the box with OmniDB 2.6.0

You are right: Firefox 59 works after I added a manual exception exception to the port 10000!

And the best: I understood the problem. :-)
Of course this might not be the best user experience.... I can deal with this now, but others might run into the same problem.. (Maybe you can provocate a Firefox warning for the websocket-port somehow by including a tiny-broken-pixel-image loading from https://:10000/ ?? just an idea)

Thanks a lot and I will try to test the "external_websocket_port" soon.

Steve

**Comment by rafaelthca** [comment]
> (Maybe you can provocate a Firefox warning for the websocket-port somehow by including a tiny-broken-pixel-image loading from https://:10000/ ?? just an idea)

That is actually a great idea. Only possible because the websocket server can also serve https requests, will do some tests here. Thank you for the suggestion.

**Comment by wind39** [comment]
Hi @alfonx , @rafaelthca ,

Reviewing this, I think perhaps we can add a note for Firefox users in OmniDB documentation.

What is your opinion?

Regards,

William

**Comment by alfonx** [comment]
Dear William. Great idea. I haven't used OmniDB for a while. But adding troubleshooting-Information sounds great. +1

**Comment by iacsvrn** [comment]
I want up this problem.

OmniDB is a perfect product, but I can`t create normal work by http <--> websocket.

At official documentation [https://www.omnidb.org/en/documentation-en/19-deploying-omnidb-server](url) we see how to write wss at nginx, but if I have nginx as a reverse proxy ( http only ), I can`t proxy it normaly.

I try to declare /wss and /ws, but it did not lead to any success. Site is "half-connection": web is ok, but query server doesn`t available.

At omnidb server, I haven`t got installed nginx server. Only server OmniDB ( Debian 9 )

**Comment by iacsvrn** [comment]
> Hi,
> 
> I'm going to work on this. Obviously needs improvement.
> 
> @iacsvrn, can you describe your architecture? Where is OmniDB installed, if it is behind a reverse proxy. Also share all the configurations please.
> 
> Rafael.

Hello! Yep, It`s behind reverse proxy (nginx)

Architecture: nginx(http) <--> omnidb

**Comment by barthisrael** [comment]
Hello everyone,

We invite you to test **OmniDB 3.0**, which is not using websockets anymore. This kind of errors are not expected anymore.

Regards.

**Comment by iacsvrn** [comment]
> 
> 
> Hello everyone,
> 
> We invite you to test **OmniDB 3.0**, which is not using websockets anymore. This kind of errors are not expected anymore.
> 
> Regards.

Big thanks to your work =)
